### DISCUSSION TITLE
Request for Assistance with Sim2Real Transformation Using Emitter Receiver in DeepBots

### DISCUSSION THREAD

**Discussion by abdul-mannan-khan** [cyberbotics/webots#6462] [Ideas]
Hello Cyberbotics/Webots,

Thank you for sharing this amazing support WeBots.
I am wondering about Sim2Real transformations, specifically within the context of using emitter-receiver configurations in [DeepBots](https://github.com/aidudezzz). I am seeking guidance or any helpful resources that could steer me in the right direction for this endeavor.

My current project involves a Sim2Real transformation where I am attempting to bridge the gap between the simulated environment and the real-world application using a setup based on [DeepBots](https://github.com/aidudezzz). I am particularly focused on understanding and implementing the emitter-receiver functionality effectively for this transition.

As a beginner in this area, any advice, tutorials, or shared experiences would be incredibly valuable to me. I am especially interested in any best practices or common pitfalls to avoid during this process.

Thank you in advance for your time and assistance. Your insights will be greatly appreciated and will surely contribute significantly to my learning journey in this exciting field.

**Comment by ShuffleWire** [comment]
Hello,

I don't know if support is still wanted / needed, but I could help work with Webots.
What is blocking you ?
Didi you get the emitter-receiver working ?

**Reply by abdul-mannan-khan** [reply]
Thank you @ShuffleWire for your response. I got emitter-receiver working. However, the problem is that when I put this simulation in SLURM-based managed Hight Performance Computing system, I get errors (related to PROTOS). In particulary, The main problem is as follows


```
ERROR: Missing declaration for 'undefined', unknown
 node.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto':19:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto':19:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto': error: PROTO has unknown base type.
WARNING: '/user/home/yf23868/.cache/Cyberbotics/Webots/assets/9d44e7d180f038a78f09d020138c7c12caadc318': This file was created by Webots R2023b while you are using Webots R2023a revision 1. Forward compatibility may not work.
WARNING: '/user/home/yf23868/.cache/Cyberbotics/Webots/assets/a1d30de7ec4c70e2553ef0bb73393ff766e23f8f': This file was created by Webots R2023b while you are using Webots R2023a revision 1. Forward compatibility may not work.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto':45:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto':45:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto': error: PROTO has unknown base type.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto':43:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto':43:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto': error: PROTO has unknown base type.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto':19:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto':19:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto': error: PROTO has unknown base type.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto':24:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto':24:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto': error: PROTO has unknown base type.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/SimpleBuilding.proto':36:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/SimpleBuilding.proto':36:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/SimpleBuilding.proto': error: PROTO has unknown base type. 
```


if you go to that website link and that line number you will find (for example 

`https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto':24:1: error: Skipped unknown 'undefined' node or PROTO.`) problem in following code snipped:


```
  %<
    let size = fields.size.value;
    if (size.x <= 0.0 || size.y <= 0.0) {
      size = fields.size.defaultValue;
      console.error('\'size\' must contain positive values. Value reset to (' + size.x + ', ' + size.y + ').');
    }

    let texture = fields.texture.value[0];
    if (texture !== fields.texture.defaultValue[0])
      console.error('The \'texture\' field is deprecated, using the new \'appearance\' field instead.');

    let tileSize = fields.tileSize.value;
    if (tileSize.x <= 0 || tileSize.y <= 0) { // avoid a zero division and negative values
      tileSize = fields.tileSize.defaultValue;
      console.error('\'size\' must contain positive values. Value reset to (' + tileSize.x + ', ' + tileSize.y + ').');
    }
    let textureScale = {x: size.x / tileSize.x, y: size.y / tileSize.y};
  >%
```
this seems some sort code part related to Java. I have tried many things like putting these "proto" locally and not getting them from the website. But still it throws errors. Interestingly, the same simulation works locally on Linux Ubuntu 20.04 and 18.04. I have also tested it using Singularity contrainer with Linux ranging from 18.04 to 22.04 locally with compatible WeBots. However, when I put this simulation on HPC, I get the above mentioned error. 

Your kind help is highly apprecaited. We are facing this problem for very long time.

**Reply by ShuffleWire** [reply]
0) I've been working on those SLURM worker few years back, so I get to overall idea, but it's old now !
1) Could you please reformat your post, it's a bit difficult to read what is what. I think the %< %> confuse github...
2) Have your tried using the version R2023b ? I see that your file trigger the warning : `This file was created by Webots R2023b while you are using Webots R2023a revision 1` There are some change between version that could get webots in trouble...
3) Have you tried to load the most basic world file ? Like this
```
WorldInfo{}
Viewpoint{}
```
4) It could be that the script part of the proto is not evaluated correctly, and then Webots failed to parse as a regular PROTO. Are you sure your SLURM worker is capable of running javacript ? That being said, it should be bundled with webots anyway... But try to use non scripted PROTO first, and then move on.

**Reply by abdul-mannan-khan** [reply]
Thank you @ShuffleWire for your prompt response. I have reformated the post. Yes, it was confusing. Appologies for missing it. 

I have tried running very basic file with simple drone (Tutorial) on SLURM based HPC and the behavior was same. The drone was not able to detect the surface and just kept falling down. I have also tried using 2023b version and also 2023a, 2022b, 2022a and so on. This is consistent. If you can help me, I would be really grateful to you.

**Reply by ShuffleWire** [reply]
AFAIK, the Floor is not loaded at all by Webots, due to the error
```
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto':24:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto':24:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto': error: PROTO has unknown base type.
```
So we need to fix it first. Consider that if any loading fail, you won't be able to do anything usefull !

**Reply by ShuffleWire** [reply]
The useful test to make is to remove all script from PROTO node to see if it improve the situation, at least at loading time.
For that, open your world locally, and navigate to all the node that trigger error : SimpleBuilding, Floor, Parquetry
For each one, in the tree, right click on them and Select "View Generated PROTO Node"
It will open up a new windows with the PROTO without script. Of course, you lose the ability to change it, but it's fine for now. 
For each one, create a new PROTO, place those in your "protos" folder, rename them to "FloorLocal" for instance (for avoiding name clash) (and don't forget to modify also the name IN the PROTO file)
Now ensure that those file will be loaded instead, by updating your world.

At this point, your should not have any script in those object, upload that to the HPC and see if the situation improve.

I hope I'm clear xD let's me know of any issue. 

Could you also share the EXACT world that is causing you trouble ?

**Reply by abdul-mannan-khan** [reply]
[supervisor_world.wbt.txt](https://github.com/cyberbotics/webots/files/15250228/supervisor_world.wbt.txt)
Hello ShuffleWire, 

Thank you so much for your detailed response. I have tried putting all the PROTOS locally. Unfortunately, there was no difference in the error I was receiving. I have copied the error here:


```
ERROR: Skipped PROTO 'Parquetry' as it is not available at: /mnt/worlds/protos/Parquetry.proto.
ERROR: Skipped PROTO 'Building' as it is not available at: /singularity_home/SafeAIControlledDrones/protos/Building.proto.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/Floor.proto':18:36: error: Skipped unknown 'Parquetry' node or PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/SimpleBuilding.proto':241:3: error: Errors when parsing the base PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/worlds/supervisor_world_local_2024.wbt':388:14: error: Skipped unknown 'SimpleBuilding' node or PROTO.
ERROR: Missing declaration for 'undefined', unknown node.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/TexturedBackground.proto':45:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/TexturedBackground.proto':45:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/TexturedBackground.proto': error: PROTO has unknown base type.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/TexturedBackgroundLight.proto':43:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/TexturedBackgroundLight.proto':43:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/TexturedBackgroundLight.proto': error: PROTO has unknown base type.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto':19:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto':19:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: 'https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/Parquetry.proto': error: PROTO has unknown base type.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/Floor.proto':24:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/Floor.proto':24:1: error: Skipped unknown 'undefined' node or PROTO.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/Floor.proto': error: PROTO has unknown base type.
ERROR: '/singularity_home/SafeAIControlledDrones/protos/SimpleBuilding.proto':241:3: error: Errors when parsing the base PROTO.
```

I have attached my .wbt file with extension .txt as it was not possible for me to add it without it. Could you please take a look? I am stuck in this problem for a long period of time and it is a critical step in my simulation. I am more than happy to compensate for your kind efforts. 

Thank you so much @ShuffleWire for your help.

**Reply by ShuffleWire** [reply]
Ok, in this situation you need to **reduce the number of stuff** you are testing and invoking. Your world trigger many time the same king of error, so it make kind of a mess.
Lets dissect it and take it step by step.
We will test that loading very basic Proto work fine on your setup, and work our way to complex thing later.

Using this project : [test_proto.zip](https://github.com/cyberbotics/webots/files/15261620/test_proto.zip)
The world is a very basic White Floor, with nothing else.
The trick is the following : The node Floor is actually a LocalFloor, for which the definition is located in the protos directory I provide you.
This proto is almost the same that the original one, but I've removed all the scripted stuff, and without any external dependencies.
So Webots should be able to open it just fine.

**So what I want is :** 
1) Get the version of your Webots (I suspect it cause trouble) : Run the command line : `webots --version`, OR "Help/About..."
2) Open the world I send you, and copy the error log, if any.
3) Give a screenshot of Webots.

You should get something like : 
![image](https://github.com/cyberbotics/webots/assets/83014015/c04d2b1f-03f9-477c-9128-a5aee2e8d029)

If it doesn't work, try to change in the wbt AND the proto file the first line to match the version of Webots you are actually using. 
for instance it look like you are using Webots 2023a so try : 
`#VRML_SIM R2023a utf8`

Also in the wbt you are sending me the version is Webots 2022a, which is both old, and not matching with your setup, it seems.

**Be careful** when working with different version of Webots, the world file and PROTO are not always compatible.
I would **choose and enforce a given version** and stick with it Locally and Remotely.
